File: plugins/capawesome-capacitor-bluetooth-low-energy/ios/Plugin/Classes/Options/WriteCharacteristicOptions.swift
import Foundation
import CoreBluetooth
import Capacitor

@objc public class WriteCharacteristicOptions: NSObject {
    private var characteristicId: String
    private var deviceId: String
    private var serviceId: String
    private var type: CBCharacteristicWriteType
    private var value: Data

    init(characteristicId: String, deviceId: String, serviceId: String, type: String?, value: [UInt8]) {
        self.characteristicId = characteristicId
        self.deviceId = deviceId
        self.serviceId = serviceId
        self.type = WriteCharacteristicOptions.getTypeFromString(type: type)
        self.value = BluetoothLowEnergyHelper.convertJsonArrayToData(values: value)
    }

    func getCharacteristicId() -> String {
        return characteristicId
    }

    func getDeviceId() -> String {
        return deviceId
    }

    func getServiceId() -> String {
        return serviceId
    }

    func getType() -> CBCharacteristicWriteType {
        return type
    }

    func getValue() -> Data {
        return value
    }

    private static func getTypeFromString(type: String?) -> CBCharacteristicWriteType {
        switch type {
        case "withoutResponse":
            return CBCharacteristicWriteType.withoutResponse
        default:
            return CBCharacteristicWriteType.withResponse
        }
    }
}

File: plugins/capawesome-capacitor-bluetooth-low-energy/android/src/main/java/io/capawesome/capacitorjs/plugins/bluetoothle/classes/options/WriteCharacteristicOptions.java
package io.capawesome.capacitorjs.plugins.bluetoothle.classes.options;

import android.bluetooth.BluetoothGattCharacteristic;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import io.capawesome.capacitorjs.plugins.bluetoothle.BluetoothLowEnergyHelper;
import org.json.JSONException;

public class WriteCharacteristicOptions {

    @NonNull
    private String characteristicId;

    @NonNull
    private String deviceId;

    @NonNull
    private String serviceId;

    @NonNull
    private int type;

    @NonNull
    private byte[] value;

    public WriteCharacteristicOptions(
        @NonNull String characteristicId,
        @NonNull String deviceId,
        @NonNull String serviceId,
        @Nullable String type,
        @NonNull JSArray value
    ) throws JSONException {
        this.characteristicId = characteristicId;
        this.deviceId = deviceId;
        this.serviceId = serviceId;
        this.value = BluetoothLowEnergyHelper.convertJsonArrayToByteArray(value);
    }

    @NonNull
    public String getCharacteristicId() {
        return characteristicId;
    }

    @NonNull
    public String getDeviceId() {
        return deviceId;
    }

    @NonNull
    public String getServiceId() {
        return serviceId;
    }

    public int getType() {
        return type;
    }

    @NonNull
    public byte[] getValue() {
        return value;
    }

    private int getTypeFromString(String type) {
        switch (type) {
            case "withoutResponse":
                return BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE;
            default:
                return BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT;
        }
    }
}
